agent/lib_composer.c

```diff
@@ -0,0 +1,233 @@
+    zend_string* package_name = NULL;
+    zval* package_version = NULL;
+    ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL(retval), package_name,
+                                  package_version) {
+      if (NULL == package_name || NULL == package_version) {
+        continue;
+      }
+      if (nr_php_is_zval_non_empty_string(package_version)) {
+        nrl_verbosedebug(NRL_INSTRUMENT, "package %s, version %s",
+                         NRSAFESTR(ZSTR_VAL(package_name)),
+                         NRSAFESTR(Z_STRVAL_P(package_version)));
+        nr_txn_add_php_package_from_source(
+            NRPRG(txn), NRSAFESTR(ZSTR_VAL(package_name)),
+            NRSAFESTR(Z_STRVAL_P(package_version)),
+            NR_PHP_PACKAGE_SOURCE_COMPOSER);
+      }
+    }
+    ZEND_HASH_FOREACH_END();
+  } else {
+    nrl_verbosedebug(NRL_INSTRUMENT, "%s - installed packages is not an array",
+                     __func__);
+  }
+  zval_dtor(&retval);
+}
+
+static char* nr_execute_handle_autoload_composer_get_vendor_path(
+    const char* filename) {
+  char* vendor_path = NULL;  // result of dirname(filename)
+  char* cp = NULL;
+
+  // vendor_path = dirname(filename):
+  // 1. copy filename to vendor_path
+  vendor_path = nr_strdup(filename);
+  // 2. // find last occurence of '/' in vendor_path
+  cp = nr_strrchr(vendor_path, '/');
+  // 3. replace '/' with '\0' to get the directory path
+  *cp = '\0';
+
+  return vendor_path;
+}
+
+static bool nr_execute_handle_autoload_composer_file_exists(
+    const char* vendor_path,
+    const char* filename) {
+  char* composer_magic_file = NULL;  // vendor_path + filename
+  bool file_exists = false;
+
+  composer_magic_file = nr_formatf("%s/%s", vendor_path, filename);
+  if (0 == nr_access(composer_magic_file, F_OK | R_OK)) {
+    file_exists = true;
+  }
+  nr_free(composer_magic_file);
+  return file_exists;
+}
+
+void nr_composer_handle_autoload(const char* filename) {
+// Composer signature file"
+#define COMPOSER_MAGIC_FILE_1 "composer/autoload_real.php"
+#define COMPOSER_MAGIC_FILE_1_LEN (sizeof(COMPOSER_MAGIC_FILE_1) - 1)
+// Composer runtime API files:
+#define COMPOSER_MAGIC_FILE_2 "composer/InstalledVersions.php"
+#define COMPOSER_MAGIC_FILE_2_LEN (sizeof(COMPOSER_MAGIC_FILE_2) - 1)
+#define COMPOSER_MAGIC_FILE_3 "composer/installed.php"
+#define COMPOSER_MAGIC_FILE_3_LEN (sizeof(COMPOSER_MAGIC_FILE_3) - 1)
+  char* vendor_path = NULL;  // result of dirname(filename)
+
+  vendor_path = nr_execute_handle_autoload_composer_get_vendor_path(filename);
+  if (NULL == vendor_path) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "unable to get vendor path from '%s'",
+                     filename);
+    return;
+  }
+
+  if (!nr_execute_handle_autoload_composer_file_exists(vendor_path,
+                                                       COMPOSER_MAGIC_FILE_1)) {
+    nrl_verbosedebug(NRL_FRAMEWORK, "'%s' not found in '%s'",
+                     COMPOSER_MAGIC_FILE_1, vendor_path);
+    return;
```

Comment:
vendor_path needs to be freed

Reply:
Fixed in fd3f639501a84d36e23fd773ee8ad166176b8330.